framework/src/modules/interoperability/base_interoperability_module.ts

```diff
@@ -50,4 +58,275 @@ export abstract class BaseInteroperabilityModule extends BaseInteroperableModule
 		this.interoperableCCCommands.set(module.name, module.crossChainCommand);
 	}
+
+	public async initGenesisState(ctx: GenesisBlockExecuteContext): Promise<void> {
+		const assetBytes = ctx.assets.getAsset(MODULE_NAME_INTEROPERABILITY);
+		if (!assetBytes) {
+			return;
+		}
+		const mainchainID = getMainchainID(ctx.chainID);
+		const mainchainTokenID = getMainchainTokenID(ctx.chainID);
+
+		const genesisStore = codec.decode<GenesisInteroperability>(
+			genesisInteroperabilitySchema,
+			assetBytes,
+		);
+		validator.validate<GenesisInteroperability>(genesisInteroperabilitySchema, genesisStore);
+
+		const outboxRootStoreKeySet = new dataStructures.BufferSet();
+		const outboxRootStore = this.stores.get(OutboxRootStore);
+		for (const outboxRootData of genesisStore.outboxRootSubstore) {
+			if (outboxRootStoreKeySet.has(outboxRootData.storeKey)) {
+				throw new Error(
+					`Outbox root store key ${outboxRootData.storeKey.toString('hex')} is duplicated.`,
+				);
+			}
+			outboxRootStoreKeySet.add(outboxRootData.storeKey);
+			await outboxRootStore.set(ctx, outboxRootData.storeKey, outboxRootData.storeValue);
+		}
+
+		const channelDataStoreKeySet = new dataStructures.BufferSet();
+		const channelDataStore = this.stores.get(ChannelDataStore);
+		for (const channelData of genesisStore.channelDataSubstore) {
+			if (channelDataStoreKeySet.has(channelData.storeKey)) {
+				throw new Error(
+					`Channel data store key ${channelData.storeKey.toString('hex')} is duplicated.`,
+				);
+			}
+			channelDataStoreKeySet.add(channelData.storeKey);
+
+			const channel = channelData.storeValue;
+			const [chainID] = splitTokenID(channel.messageFeeTokenID);
+
+			if (
+				!channel.messageFeeTokenID.equals(mainchainTokenID) && // corresponding to the LSK token
+				!chainID.equals(channelData.storeKey) && // Token.getChainID(channel.messageFeeTokenID) must be equal to channelData.storeKey
+				!chainID.equals(ctx.chainID) // the message fee token must be a native token of either chains
+			) {
+				throw new Error(
+					`messageFeeTokenID corresponding to the channel data store key ${channelData.storeKey.toString(
+						'hex',
+					)} is not valid.`,
+				);
+			}
+
+			await channelDataStore.set(ctx, channelData.storeKey, channelData.storeValue);
+		}
+
+		const chainValidatorsStoreKeySet = new dataStructures.BufferSet();
+		const chainValidatorsStore = this.stores.get(ChainValidatorsStore);
+		for (const chainValidators of genesisStore.chainValidatorsSubstore) {
+			if (chainValidatorsStoreKeySet.has(chainValidators.storeKey)) {
+				throw new Error(
+					`Chain validators store key ${chainValidators.storeKey.toString('hex')} is duplicated.`,
+				);
+			}
+			chainValidatorsStoreKeySet.add(chainValidators.storeKey);
+
+			const { activeValidators, certificateThreshold } = chainValidators.storeValue;
+
+			let totalWeight = BigInt(0);
+			for (let j = 0; j < activeValidators.length; j += 1) {
+				const activeValidator = activeValidators[j];
```

Comment:
I think `activeValidators must have at least 1 element and at most MAX_NUM_VALIDATORS elements.` validation is missing for chainValidators

Reply:
number of activeValidators is checked in `chainValidatorsSchema` with minItems and maxItems